Classes/FairPlayDRM/FPSAssetLoaderDelegate.swift

```diff
@@ -0,0 +1,174 @@
+    }
+    
+    func prepareAndSendContentKeyRequest(resourceLoadingRequest: AVAssetResourceLoadingRequest) {
+        
+        guard let assetId = resourceLoadingRequest.request.url?.host else {
+            PKLog.error("No asset id")
+            return
+        }
+        
+        // Check if this reuqest is the result of a potential AVAssetDownloadTask.
+        if #available(iOS 10.0, *), shouldPersist {
+            if let cir = resourceLoadingRequest.contentInformationRequest {
+                cir.contentType = AVStreamingKeyDeliveryPersistentContentKeyType
+            }
+        }
+        
+        guard let helper = FPSLicenseHelper.init(assetId: assetId, params: self.drmData, 
+                                                 dataStore: self.storage, forceDownload: self.forceDownload) else { return }
+        
+        helper.handleLicenseRequest(FPSResourceLoadingKeyRequest(resourceLoadingRequest)) { (error) in
+            self.done?(error)
+        }
+    }
+
+    func shouldLoadOrRenewRequestedResource(resourceLoadingRequest: AVAssetResourceLoadingRequest) -> Bool {
+        
+        guard let url = resourceLoadingRequest.request.url else {
+            return false
+        }
+        
+        // AssetLoaderDelegate only should handle FPS Content Key requests.
+        if url.scheme != FPSAssetLoaderDelegate.customScheme {
+            return false
+        }
+        
+        FPSAssetLoaderDelegate.resourceLoadingRequestQueue.async {
+            self.prepareAndSendContentKeyRequest(resourceLoadingRequest: resourceLoadingRequest)
+        }
+        
+        return true
+    }
+}
+
+//MARK:- AVAssetResourceLoaderDelegate protocol methods extension
+extension FPSAssetLoaderDelegate: AVAssetResourceLoaderDelegate {
+    
+    func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForLoadingOfRequestedResource loadingRequest: AVAssetResourceLoadingRequest) -> Bool {
+        
+        PKLog.trace("\(#function) was called in AssetLoaderDelegate with loadingRequest: \(loadingRequest)")
+        
+        return shouldLoadOrRenewRequestedResource(resourceLoadingRequest: loadingRequest)
+    }
+    
+    func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForRenewalOfRequestedResource renewalRequest: AVAssetResourceRenewalRequest) -> Bool {
+        
+        PKLog.trace("\(#function) was called in AssetLoaderDelegate with renewalRequest: \(renewalRequest)")
+        
+        return shouldLoadOrRenewRequestedResource(resourceLoadingRequest: renewalRequest)
+    }
+}
+
+class FPSResourceLoadingKeyRequest: FPSLicenseRequest {
+    let request: AVAssetResourceLoadingRequest
+    init(_ request: AVAssetResourceLoadingRequest) {
+        self.request = request
+    }
+    
+    func getSPC(cert: Data, id: String, shouldPersist: Bool, callback: @escaping (Data?, Error?) -> Void) {
+        var options: [String: AnyObject]? = nil
+        
+        if #available(iOS 10.0, *), shouldPersist {
```

Comment:
AVAssetResourceLoadingRequestStreamingContentKeyRequestRequiresPersistentKey - is available from iOS 9+ from documentation... why is it checking from iOS 10.
What happens in tvOS?

Reply:
iOS 9 doesn't support offline FairPlay, even though the docs say it does. 
tvOS doesn't support offline anything - see that AVAssetDownloadTask is prohibited in tvOS.